packages/gatsby/index.d.ts

```diff
@@ -29,6 +29,78 @@ export const parsePath: (path: string) => WindowLocation
 
 export const prefetchPathname: (path: string) => void
 
+/** 
+ * A props object for adding type safety to your Gatsby pages, can be
+ * extended with both the query response shape, and the page context.
+ * 
+ * @example 
+ * // When typing a default page from the ./pages dir
+ * 
+ * import {PageProps} from "gatsby"
+ * export default (props: PageProps) => {
+ * 
+ * @example
+ * // When adding types for both pageContext and GraphQL query data
+ * 
+ * import {PageProps} from "gatsby"
+ * 
+ * type IndexQueryProps = { downloadCount: number }
+ * type LocaleLookUpInfo = { translationStrings: any } & { langKey: string, slug: string }
+ * type IndexPageProps = PageProps<IndexPageProps, LocaleLookUpInfo>
+ * 
+ * export default (props: IndexProps) => {
+ *   ..
+ */
+export type PageProps<DataType = object, PageContextType = { langKey: string, slug: string }> = {
```

Comment:
We don't want any default here, it's up to the user to give us a pageContext or not. Or am I missing something?

```suggestion
export type PageProps<DataType = object, PageContextType = object> = {
```

Reply:
That default object I provided here is what you get when you place a file in the `pages` dir. Where you basically don't run `createPage` yourself - which is about as default-y as you can get IMO. 

This means you only need to override it if you have a `createPage` with a custom context, changing it would mean you _always_ have to override it.

Reply:
Both `langKey` and `slug` doesn't come from gatsby core - it must be some i18n plugin adding those:

We use whatever some user (or plugin) give us in core (default to empty object if it was undefined/falsy):
https://github.com/gatsbyjs/gatsby/blob/f5acf522feb48ad211e509a788e870effdddbe20/packages/gatsby/src/redux/actions/public.js#L376-L377

The pages from `src/pages` are having `undefined` context:
https://github.com/gatsbyjs/gatsby/blob/f5acf522feb48ad211e509a788e870effdddbe20/packages/gatsby-plugin-page-creator/src/gatsby-node.js#L96-L101

The `langKey` and `slug` seems to be provided by `gatsby-plugin-i18n` (or different - that's first i18n plugin I checked, but others might use similar convention):
https://github.com/angeloocana/gatsby-plugin-i18n/blob/683c3c3c935629af78673d23ca9d02dbf7216397/packages/gatsby-plugin-i18n/src/onCreatePage.js#L5-L11

Reply:
That fits - nice digging @pieh thanks!

OK, then @wardpeet's suggestion is the best option then 👍